agent/dockerclient/dockerapi/docker_events_buffer.go

```diff
@@ -51,15 +52,32 @@ func NewInfiniteBuffer() *InfiniteBuffer {
 }
 
 // StartListening starts reading from the input channel and writes to the buffer
-// TODO: wire in ctx to stop listening
-func (buffer *InfiniteBuffer) StartListening(events chan *docker.APIEvents) {
-	for event := range events {
-		go buffer.CopyEvents(event)
+// When context is cancelled, stop listening
+func (buffer *InfiniteBuffer) StartListening(ctx context.Context, eventChan <-chan events.Message) {
+	for {
+		select {
+		// If context is cancelled, return
+		case <-ctx.Done():
+			select {
+			// Ensure all events are read in the channel before we return
+			case event := <-eventChan:
```

Comment:
I don't think this is necessary, as this only reads one more event from the channel, also in the previous code the new channel was opened first and then the context was canceled.